lxd/rbac/server.go

```diff
@@ -0,0 +1,427 @@
+package rbac
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/juju/persistent-cookiejar"
+	"gopkg.in/macaroon-bakery.v2/bakery"
+	"gopkg.in/macaroon-bakery.v2/httpbakery"
+	"gopkg.in/macaroon-bakery.v2/httpbakery/agent"
+
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/logger"
+)
+
+// UnexpectedSyncIDError is returned when the RBAC server received an unexpected sync ID.
+type UnexpectedSyncIDError struct{}
+
+// Error implements error.Error.
+func (e *UnexpectedSyncIDError) Error() string {
+	return "Unexpected sync ID"
+}
+
+type rbacResource struct {
+	Identifier string `json:"identifier"`
+	Name       string `json:"name"`
+}
+
+type rbacResourcePost struct {
+	LastSyncID *string        `json:"last-sync-id"`
+	Updates    []rbacResource `json:"updates,omitempty"`
+	Removals   []string       `json:"removals,omitempty"`
+}
+
+type rbacResourcePostResponse struct {
+	SyncID string `json:"sync-id"`
+}
+
+type rbacStatus struct {
+	LastChange time.Time `json:"last-change"`
+}
+
+// Server represents an RBAC server.
+type Server struct {
+	apiURL string
+	apiKey string
+
+	lastSyncID string
+	client     *httpbakery.Client
+	lastChange time.Time
+	statusDone chan int
+
+	resources   map[string]string // Maps name to identifier
+	permissions map[string]map[string][]string
+
+	permissionsLock *sync.Mutex
+}
+
+// NewServer returns a new RBAC server instance.
+func NewServer(apiURL string, apiKey string, agentAuthURL string, agentUsername string, agentPrivateKey string, agentPublicKey string) (*Server, error) {
+	r := Server{
+		apiURL:          apiURL,
+		apiKey:          apiKey,
+		lastSyncID:      "",
+		lastChange:      time.Time{},
+		resources:       make(map[string]string),
+		permissions:     make(map[string]map[string][]string),
+		permissionsLock: &sync.Mutex{},
+	}
+
+	var keyPair bakery.KeyPair
+
+	keyPair.Private.UnmarshalText([]byte(agentPrivateKey))
+	keyPair.Public.UnmarshalText([]byte(agentPublicKey))
+
+	r.client = httpbakery.NewClient()
+	authInfo := agent.AuthInfo{
+		Key: &keyPair,
+		Agents: []agent.Agent{
+			{
+				URL:      agentAuthURL,
+				Username: agentUsername,
+			},
+		},
+	}
+
+	err := agent.SetUpAuth(r.client, &authInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	r.client.Client.Jar, err = cookiejar.New(&cookiejar.Options{
+		Filename: cookiejar.DefaultCookieFile(),
```

Comment:
We might also want to change this. `DefaultCookieFile()` points to `filepath.Join(homeDir(), ".go-cookies")`.